src/ImageSharp/Formats/ImageDecoder.cs

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using System.IO;
+using SixLabors.ImageSharp.IO;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.Processing;
+
+namespace SixLabors.ImageSharp.Formats;
+
+/// <summary>
+/// Acts as a base class for image decoders.
+/// Types that inherit this decoder are required to implement cancellable synchronous decoding operations only.
+/// </summary>
+public abstract class ImageDecoder : IImageDecoder
+{
+    /// <inheritdoc/>
+    public Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream)
+        where TPixel : unmanaged, IPixel<TPixel>
+        => WithSeekableStream(
+              options,
+              stream,
+              s => this.Decode<TPixel>(options, s, default));
+
+    /// <inheritdoc/>
+    public Image Decode(DecoderOptions options, Stream stream)
+        => WithSeekableStream(
+              options,
+              stream,
+              s => this.Decode(options, s, default));
+
+    /// <inheritdoc/>
+    public Task<Image<TPixel>> DecodeAsync<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken = default)
```

Comment:
Custom decoders may not need to enforce rewrapping into `ChunkedMemoryStream`, especially TiffLibrary's pure async decoder.

Is our recommendation to implement `IImageDecoder` directly then? Would it be safe?

Reply:
But then there is no way to `ScaleToTargetSize`.

We can make these methods virtual, but then there is no need to override the abstract ones. We can turn the abstract methods to virtual ones throwing `NotImplementedException`, but it would become confusing.

Any better ideas?

Reply:
Yeah, they would implement `IImageDecoder` directly since we don't expose any async virtuals in `ImageDecoder`

Reply:
Re ScaleToTargetSize. My concern with moving it to a shared utility was that someone might end up implementing a per-scanline scaling operation.

It's tricky.

Reply:
> per-scanline scaling operation.

If we have the time, we definitely need to do that for 3.0 at least for JPEG+PNG. The implementation codein  `***DecoderCore` classes will be practically feeding scanlines to [`ResizeWorker`](https://github.com/SixLabors/ImageSharp/blob/7b403897dfee6433d7c39feb6fdef1f4dfe0949b/src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeWorker.cs) or whatever it gets refactored to. Not sure if that infrastructure will be mature enough to be exposed as a public API, that's a secondary priority IMO.